### ISSUE TITLE
Breaks if node runs with "--experimental-modules"

### ISSUE THREAD

**Issue by alqu** [jeffijoe/awilix-router-core#6]
Hey,

thanks for this nice library!
Unfortunately this does not run out-of-the box (a.k.a. without babel) with Node and the "experimental modules" feature switched on.
It breaks on the following line - where you require a module instead of importing it.
https://github.com/jeffijoe/awilix-router-core/blob/5b6fd8875d524ae3bb5c4768f3329aefb065846b/src/find-controllers.ts#L24

When I switch the whole project to esm with this line in package.json:
```json
"type": "module",
```

Then everything works... except the line I mentioned above.

But changing that line to
```js
import(path).then(required => { ... }
```
Seems to work.
But since this router should be able to run in both worlds - modules and commonjs - there need to be some tuning here and there.

Do you have any experience with this?

**Comment by jeffijoe** [comment]
I haven't used native support for modules yet, so I don't have any experience with it.

What is breaking specifically? What's the error? I would imagine backwards-compatibility with `require` is a priority to Node.

**Comment by alqu** [comment]
Hm, switching node into the "I am more modern" mode removes require (as a function). What then works is the dynamic import statement above.... of course asynchrous.
Just try it out with
- adding the line ```json
"type": "module",
```
- and starting your app with e.g.
```js
nodemon --experimental-modules bin/www.js"
```

That should expose the problem.

**Comment by alqu** [comment]
...the error is:
```
  12:32:21:808 ERROR Error while starting up server (bin/www.js:11)
    →  myproject/backend/routes/my-api.js:1
    →  import awilixKoa from 'awilix-koa'
```

**Comment by jeffijoe** [comment]
Not sure how to tackle this one given we currently rely on the synchronous nature of `require`. I want to avoid making the outer API asynchronous, it would be unnatural to have to do `app.use(await loadControllers(...))`.

**Comment by alqu** [comment]
I forgot to mention: the rest of the app works OK with import / export statements. Only the part where the loader tries to require / import the controllers breaks.

Asynchronicity is only required if you want to dynamically import something. And this is the case. For some reason `import bla from 'string_literal'` is limited to, well, a string literal.

**Comment by jeffijoe** [comment]
For the time being, you can use `app.use(controller(yourController))` to avoid going down the require path.

**Comment by alqu** [comment]
Hm. I like the loader-approach. What I do instead:
I run nodeJS in commonJS-mode (which is the default) with `"type": "commonjs"` in `package.json`, switch some directories and files into module mode, but especially all the routes and services stay "old-school" in commonJS-style. This works.

**Comment by alqu** [comment]
Sounds like a plan.

Or... do not use experimental modules at all for now.

**Comment by armpogart** [comment]
@jeffijoe Why is this issue closed? I mean, is there any plans to support ESM modules? Or is PR welcome?

**Comment by jeffijoe** [comment]
@armpogart PR is welcome, as long as it maintains backwards compatibility with the current API and with `require`. In other words, we shouldn't have to use `app.use(await loadControllers(..))`.

**Comment by axelrindle** [comment]
> @armpogart PR is welcome, as long as it maintains backwards compatibility with the current API and with `require`. In other words, we shouldn't have to use `app.use(await loadControllers(..))`.

Usage of dynamic import [requires the usage of Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import) and async/await. I don't see any other way to accomplish full ESM support.

**Comment by Worthaboutapig** [comment]
Is there any reason you can't do the same thing you've done here:

[load-modules->opts.esModules](https://github.com/jeffijoe/awilix/blob/37010cb0589ed6dec55022e2e5e098c69cbe2cdc/src/load-modules.ts#L118)

Add an esModule flag that uses a dynamic import instead?  Seems trivial, backward-compatible and consistent with how you've handled it elsewhere.

I'd have thought supporting modules was pretty-much necessary these days. I certainly need it ;)

Happy to add a PR.

**Comment by jeffijoe** [comment]
I've considered it - it would be slightly different since ideally, you wouldn't want to `app.use(await loadControllers())`, that is require the user to make their entire app setup async.

**Comment by jeffijoe** [comment]
I was thinking perhaps `loadControllers` in framework packages such as `awilix-koa` could be implemented something like

```js
export function loadControllers(opts) {
  const modulesPromise = loadTheModulesAsync()

  return async function middleware(ctx, next) {
    const modules = await modulesPromise
  }
}
```

This would make it so the consumer wouldn't have to make the caller of `loadControllers` async. Of course one downside here is any error would be an uncaught promise rejection.

What do you think?

**Comment by Worthaboutapig** [comment]
After trying around various ideas, I have this working in the way you suggest- the problem was ensuring the router middleware was run, but only after the `modulesPromise` resolves. So, the router middleware is called via `use` within the `loadControllers` middleware.

Yes, a rejection from the `modulesPromise` would not fail on initialising the middleware, but would be part of the middleware stack and would happen in the same way as any other rejection within a middleware would behave.

I guess that's the tradeoff- make the rejection happen in the initialisation (my preference), or defer it until it's resolved within the middleware.

Facts:
By `await`ing you have the advantage in failing fast at app startup, but by deferring it to the middleware stack you'll only find out about the rejection when the route is called.
Also, `await`ing within the middleware means that the resolved promise is checked on every call of the router, which is an unnecessary inefficiency.

Opinion:
Personally I don't see an issue with `await`ing `loadControllers`, you still get an array of middleware as you'd expect, which you then pass to `use`.
And I don't think requiring a top-level async function would be unexpected either these days, it's pretty normal for me.

---
That's all assuming I'm understanding you correctly and there isn't another way of doing it to avoid the issues with a deferred promise in the middleware that I can't think of.

This all depends on the `awilix-router-core` PR obviously, so when that's approved I can share a PR here and see what you think.

**Comment by jeffijoe** [comment]
Yes, you've convinced me - we'll apply the same trick with the `ESM` generic for `loadControllers` in the framework packages, thus requiring `await` when passing `esModules: true`

**Comment by Worthaboutapig** [comment]
Fixed and supported in v2.1.0
